### PR TITLE
Reconstruct the metric folder and Refactoring metric related things

### PR DESCRIPTION
**Pull request by sungmanc**
### Summary

This PR introduces,
* Move the location of `algo/metrics` --> `core/metrics`.
* Move `HLabelAccuracy` class into single accuracy.py file.
* Change the default parameter of `HLabelAccuracy` to reduce the complexity of the `configure_metric` function.
* Assign default value for all `LitModule` including cls, det, ... by using `lambda` to represent the `partial`
* Add `configure_metric`, `on_validation_start`, `on_test_start` functions to construct the `self.metric` at the `validation` and `test` phase.
* Add proper type-checking at the `validation and test related step` to avoid the mypy issues.

Discussion References:
https://github.com/openvinotoolkit/training_extensions/pull/2962#discussion_r1498619858
https://github.com/openvinotoolkit/training_extensions/pull/2962#discussion_r1498687069

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

**Comment by sungmanc** [comment]
Still I should resolve the issue of the HLable Classification, Please review other parts.

File: src/otx/core/metrics/__init__.py [hunk 1]
```diff
@@ -5,7 +5,10 @@
 
 from typing import Callable
 
-from lightning.pytorch.cli import ArgsType
 from torchmetrics import Metric
 
-MetricCallable = Callable[[ArgsType], Metric]
+from .accuracy import CustomAccuracy, HLabelAccuracy
+
+MetricCallable = Callable[[], Metric]
```
Comment (vinnamkim) [inline src/otx/core/metrics/__init__.py]:
```suggestion
MetricCallable = Callable[[], Metric] | Callable[[int], Metric]
```

File: src/otx/core/model/module/classification.py [hunk 2]
```diff
@@ -43,7 +46,7 @@ def __init__(
         torch_compile: bool,
         optimizer: list[OptimizerCallable] | OptimizerCallable = lambda p: torch.optim.SGD(p, lr=0.01),
         scheduler: list[LRSchedulerCallable] | LRSchedulerCallable = torch.optim.lr_scheduler.ConstantLR,
-        metric: MetricCallable | None = None,
+        metric: MetricCallable = lambda: Accuracy(task="multiclass"),
```
Comment (vinnamkim) [inline src/otx/core/model/module/classification.py]: Is this working without adding `num_classes`?
```suggestion
        metric: MetricCallable = lambda num_classes: Accuracy(task="multiclass", num_classes=num_classes),
```

File: src/otx/core/model/module/base.py [hunk 3]
```diff
@@ -37,15 +38,14 @@ def __init__(
...
+        metric: MetricCallable = lambda: Metric(),
     ):
         super().__init__()
 
         self.model = otx_model
         self.optimizer = optimizer
         self.scheduler = scheduler
         self.torch_compile = torch_compile
-
         self.metric = metric
```
Comment (vinnamkim) [inline src/otx/core/model/module/base.py]:
```suggestion
        self.metric_callable = metric
```

File: src/otx/core/metrics/accuracy.py [hunk 4]
```diff
@@ -1,16 +1,35 @@
...
+        self.average = average
+
+    def update(self, preds: Tensor, target: Tensor) -> None:
+        """Update state with predictions and targets."""
+
+    def compute(self) -> Tensor:
+        """Compute the metric."""
+
+    def _compute_unnormalized_confusion_metrices_from_preds(self, preds: Tensor) -> None:
+        """Compute an (unnormalized) confusion matriix for every label group."""
```
Comment (vinnamkim) [inline src/otx/core/metrics/accuracy.py]: Why is this class necessary?

File: src/otx/core/model/module/classification.py [hunk 5]
```diff
@@ -128,7 +145,7 @@ def __init__(
         torch_compile: bool,
         optimizer: list[OptimizerCallable] | OptimizerCallable = lambda p: torch.optim.SGD(p, lr=0.01),
         scheduler: list[LRSchedulerCallable] | LRSchedulerCallable = torch.optim.lr_scheduler.ConstantLR,
-        metric: MetricCallable | None = None,
+        metric: MetricCallable = lambda: MultilabelAccuracy(num_labels=1),
```
Comment (vinnamkim) [inline src/otx/core/model/module/classification.py]:
```suggestion
        metric: MetricCallable = lambda num_classes: Accuracy(task="multilabel", num_classes=num_classes),
```

File: src/otx/core/model/module/classification.py [hunk 6]
```diff
@@ -53,18 +56,32 @@ def __init__(
             metric=metric,
         )
 
-        if metric:
-            sig = inspect.signature(metric)
+    def configure_metric(self) -> None:
+        """Configure the metric."""
+        if isinstance(self.metric, partial):
```
Comment (vinnamkim) [inline src/otx/core/model/module/classification.py]:
```suggestion
    def configure_metric(self) -> None:
        """Configure the metric."""
        num_classes_augmented_params = {
             param.default if name != "num_classes" else self.model.num_classes
             for name, param in inspect.signature(self.metric_callable).parameters.items()
             if name != "kwargs"
        }
        self.metric = self.metric_callable(**num_classes_augmented_params)
        return
```

In addition, can it be the base function? Why do we need this function for every derived class?

File: src/otx/engine/engine.py [hunk 7]
```diff
@@ -743,10 +743,14 @@ def _build_lightning_module(
...
-        )
+
+        lightning_kwargs = {
+            "otx_model": model,
+            "optimizer": optimizer,
+            "scheduler": scheduler,
+            "torch_compile": False,
+        }
+        if metric:
+            lightning_kwargs["metric"] = metric
```
Comment (vinnamkim) [inline src/otx/engine/engine.py:758]: You should change typing at L730 in this file to MetricCallable

File: src/otx/core/metrics/accuracy.py (hunk 4 above)
Comment (sungmanc) [reply src/otx/core/metrics/accuracy.py]: This could be removed. It the skeleton code for the future work.

Comment (vinnamkim) [reply src/otx/core/metrics/accuracy.py]: Why not the future work just deriving from `Metric`?

File: src/otx/core/model/module/classification.py (hunk 6 above)
Comment (sungmanc) [reply src/otx/core/model/module/classification.py]: Some metrics (Dice, HLabelAccuracy) should take the different `param_dict` for successfully instantiation. However, I think we can specify only those cases. I'll move this at the base

File: src/otx/core/model/module/classification.py (hunk 2 above)
Comment (sungmanc) [reply src/otx/core/model/module/classification.py]: Yes, it works well since it's not mandatory.

File: src/otx/core/metrics/__init__.py (hunk 1 above)
Comment (sungmanc) [reply src/otx/core/metrics/__init__.py]: Done, https://github.com/openvinotoolkit/training_extensions/pull/2964/commits/3331c3032f6daad27eb5088f951abb1133a34d8b. Just quriousity, is it needed for the num_classes, right ?

File: src/otx/core/model/module/base.py (hunk 3 above)
Comment (sungmanc) [reply src/otx/core/model/module/base.py]: Done, https://github.com/openvinotoolkit/training_extensions/pull/2964/commits/3331c3032f6daad27eb5088f951abb1133a34d8b

File: src/otx/core/metrics/accuracy.py (hunk 4 above)
Comment (sungmanc) [reply src/otx/core/metrics/accuracy.py]: I removed this lines at this PR, https://github.com/openvinotoolkit/training_extensions/pull/2964/commits/3331c3032f6daad27eb5088f951abb1133a34d8b. What I meant is just for the OTX SDK metric

File: src/otx/core/model/module/classification.py (hunk 5 above)
Comment (sungmanc) [reply src/otx/core/model/module/classification.py]: Done, https://github.com/openvinotoolkit/training_extensions/pull/2964/commits/5a50299ddc614f9552abc2f787acd7a2c35c9c4a

File: src/otx/core/model/module/classification.py (hunk 2 above)
Comment (sungmanc) [reply src/otx/core/model/module/classification.py]: BTW, I edited this -> https://github.com/openvinotoolkit/training_extensions/pull/2964/commits/5a50299ddc614f9552abc2f787acd7a2c35c9c4a

File: src/otx/core/model/module/classification.py (hunk 6 above)
Comment (sungmanc) [reply src/otx/core/model/module/classification.py]: Moved this function to base and leave the functions at the segmentation, h-label classification which needs the special handling. https://github.com/openvinotoolkit/training_extensions/pull/2964/commits/3331c3032f6daad27eb5088f951abb1133a34d8b

File: src/otx/engine/engine.py (hunk 7 above)
Comment (sungmanc) [reply src/otx/engine/engine.py:758]: Done, https://github.com/openvinotoolkit/training_extensions/pull/2964/commits/3331c3032f6daad27eb5088f951abb1133a34d8b

**Review by harimkang — APPROVED** [review]

**Review by jaegukhyun — APPROVED** [review]

File: src/otx/core/metrics/__init__.py (hunk 1 above)
Comment (vinnamkim) [reply src/otx/core/metrics/__init__.py]: Yes

**Review by vinnamkim — APPROVED** [review]
